crates/mun_codegen/src/value/array_value.rs

```diff
@@ -0,0 +1,196 @@
+};
+use inkwell::types::{BasicType, PointerType};
+use inkwell::AddressSpace;
+
+impl<T: ConcreteValueType> ConcreteValueType for [T] {
+    type Value = inkwell::values::ArrayValue;
+}
+
+impl<T: PointerValueType> PointerValueType for [T] {
+    fn get_ptr_type(context: &IrTypeContext, address_space: Option<AddressSpace>) -> PointerType {
+        T::get_ptr_type(context, address_space)
+    }
+}
+
+impl<I, T: AddressableType<I>> AddressableType<I> for [T] {}
+impl<I, T: AddressableType<I>> AddressableType<[I]> for [T] {}
+
+macro_rules! impl_array(
+    ($($size:expr),+) => {
+        $(
+            impl<T: SizedValueType> ConcreteValueType for [T; $size] {
+                type Value = inkwell::values::ArrayValue;
+            }
+
+            impl<T: SizedValueType> SizedValueType for [T; $size]
+            where
+                <<T as ConcreteValueType>::Value as ValueType>::Type: ConstArrayType,
+            {
+                fn get_ir_type(context: &IrTypeContext) -> inkwell::types::ArrayType {
+                    T::get_ir_type(context).array_type($size)
+                }
+            }
+
+            impl<T: PointerValueType + SizedValueType> PointerValueType for [T; $size] {
+                fn get_ptr_type(context: &IrTypeContext, address_space: Option<AddressSpace>) -> PointerType {
+                    T::get_ptr_type(context, address_space)
+                }
+            }
+
+            impl<E: SizedValueType, T: AsValue<E>> AsValue<[E; $size]> for [T; $size]
+            where
+                E::Value: ConstArrayValue
+            {
+                fn as_value(&self, context: &IrValueContext) -> Value<[E; $size]> {
+                    let element_ir_type = E::get_ir_type(context.type_context);
+                    let values: [E::Value; $size] =
+                        array_init::from_iter(self.iter().map(|e| e.as_value(context).value))
+                            .expect("unable to construct sized array");
+                    let value = E::Value::const_array(values.as_ref(), &element_ir_type);
+                    Value::from_raw(value)
+                }
+            }
+
+            impl<E: SizedValueType, T: AsValue<E>> AsValue<[E]> for [T; $size]
+                where
+                    E::Value: ConstArrayValue
+            {
+                fn as_value(&self, context: &IrValueContext) -> Value<[E]> {
+                    let element_ir_type = E::get_ir_type(context.type_context);
+                    let values: [E::Value; $size] =
+                        array_init::from_iter(self.iter().map(|e| e.as_value(context).value))
+                            .expect("unable to construct sized array");
+                    let value = E::Value::const_array(values.as_ref(), &element_ir_type);
+                    Value::from_raw(value)
+                }
+            }
+        )+
+    }
+);
+
+impl_array!(
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 20, 24, 32, 36, 0x40, 0x80, 0x100
+);
```

Comment:
Does this imply that we only support arrays of particular sizes? If so, should we restrict it to the sizes that we actually need right now, and if someone introduces a new array size throw a `panic!` that directs to this code line?

Reply:
Okay, the compile error is a generic error highlighting that this type is not implemented, right? Is it clear from the the compile error that someone can resolve it by editing this `impl_array!` invocation?

Reply:
No you get the error AsValue (or some other trait) is not implemented for [A; 17]. I don't know how we could improve that though??

Reply:
Hmm. I am not sure if this would work work, but I saw that you had a default trait implementation somewhere, while you also had concrete trait implementations for some specific types. Would it be possible to have a default trait implementation for all array types that has does a [`compile_error`](https://doc.rust-lang.org/std/macro.compile_error.html)?

Reply:
No because every sized array type has a different signature. Thats the problem in the first place. Rust doesnt have generic consts (yet).

Reply:
Okay. I guess just leave it as is for now then 🙂